File: tests/data/http/retry.test.ts
import { delayForAttempt, BASE_DELAY_MS, MAX_DELAY_MS } from "@data/http/retry";

/**
 * C-15 — Hub'Eau n'annonce **aucun SLA et aucun quota chiffré**, et
 * l'application n'a pas de proxy pour mutualiser la charge de sa base
 * installée. La gigue n'est donc pas un raffinement : sans elle, tous les
 * appareils réessaient à la même seconde après une panne et forment un
 * troupeau tonnant contre un service public gratuit.
 */
describe("backoff exponentiel à gigue (C-15)", () => {
  /** Gigue neutralisée, pour éprouver la base seule. */
  const sansGigue = () => 0;

  it("expose ses constantes plutôt que de les enfouir", () => {
    expect(BASE_DELAY_MS).toBe(500);
    expect(MAX_DELAY_MS).toBe(30_000);
  });

  it("double à chaque tentative", () => {
    expect(delayForAttempt(0, sansGigue)).toBe(500);
    expect(delayForAttempt(1, sansGigue)).toBe(1000);
    expect(delayForAttempt(2, sansGigue)).toBe(2000);
    expect(delayForAttempt(3, sansGigue)).toBe(4000);
  });

  it("plafonne, sans jamais déborder", () => {
    // 2^20 × 500 ms dépasserait 145 heures. Le plafond n'est pas décoratif.
    //
    // Au plafond, le délai est un INTERVALLE — [15 s, 30 s) — et non une valeur
    // unique. Le fixer à 30 000 ms pile, comme le faisait la première version,
    // annulait la gigue exactement là où elle sert : une panne longue, où toute
    // la base installée réessaie en même temps.
    expect(delayForAttempt(20, sansGigue)).toBe(15_000);
    expect(delayForAttempt(100, sansGigue)).toBe(15_000);
    expect(delayForAttempt(100, () => 0.999_999)).toBeLessThan(30_000);
  });

  it("ajoute une gigue bornée par le délai lui-même", () => {
    const gigueMax = () => 0.999_999;
    const avec = delayForAttempt(1, gigueMax);
    expect(avec).toBeGreaterThan(1000);
    expect(avec).toBeLessThanOrEqual(2000);
  });

  it("respecte le plafond même avec la gigue au maximum", () => {
    // Sans ce garde-fou, la gigue ferait dépasser le plafond de 100 %.
    expect(delayForAttempt(20, () => 0.999_999)).toBeLessThanOrEqual(30_000);
  });

  it("ne rend jamais un délai négatif ni nul", () => {
    expect(delayForAttempt(0, () => 0)).toBeGreaterThan(0);
  });

  it("étale réellement les réessais entre appareils", () => {
    // Deux appareils tirant des gigues différentes ne doivent pas retomber sur
    // la même milliseconde — c'est toute la raison d'être de la gigue.
    const appareilA = delayForAttempt(3, () => 0.1);
    const appareilB = delayForAttempt(3, () => 0.9);
    expect(appareilA).not.toBe(appareilB);
    expect(Math.abs(appareilA - appareilB)).toBeGreaterThan(1000);
  });
});

describe("la gigue survit au plafond (C-15)", () => {
  it("étale encore les tentatives une fois le plafond atteint", () => {
    // L'ancienne formule plafonnait APRÈS avoir ajouté la gigue : dès la 6e
    // tentative, tous les appareils réessayaient à 30 000 ms exactement — le
    // troupeau tonnant que la gigue existe pour empêcher, au pire moment.
    for (const attempt of [6, 8, 12]) {
      expect(delayForAttempt(attempt, () => 0)).not.toBe(delayForAttempt(attempt, () => 0.99));
    }
  });

  it("ne dépasse jamais le plafond, gigue comprise", () => {
    for (const attempt of [0, 3, 6, 10, 50]) {
      expect(delayForAttempt(attempt, () => 0.999)).toBeLessThanOrEqual(MAX_DELAY_MS);
    }
  });

  it("reste croissant jusqu'au plafond", () => {
    expect(delayForAttempt(1, () => 0)).toBeGreaterThan(delayForAttempt(0, () => 0));
    expect(delayForAttempt(3, () => 0)).toBeGreaterThan(delayForAttempt(2, () => 0));
  });
});

File: test/data/http/retry_test.dart
// Verrouille le recul exponentiel à gigue injectée (C-12) : aucun quota
// chiffré n'est publié, aucun service intermédiaire ne mutualise la charge —
// sans étalement, tous les appareils réessaient à la même seconde. Le
// plafond porte sur la base, pas sur le résultat (voir doc de
// [delayForAttempt]) : deux tests au plateau avec gigue non nulle vérifient
// que la gigue reste visible au-delà de la sixième tentative.
import 'package:flutter_test/flutter_test.dart';
import 'package:martinpecheur/data/http/retry.dart';

void main() {
  group('Constantes (C-12)', () {
    test('baseDelay vaut 500 ms et maxDelay vaut 30 s', () {
      expect(baseDelay, const Duration(milliseconds: 500));
      expect(maxDelay, const Duration(seconds: 30));
    });
  });

  group('delayForAttempt, gigue nulle', () {
    double zeroJitter() => 0.0;

    test('double à chaque tentative avant le plateau', () {
      expect(
        delayForAttempt(0, jitter: zeroJitter),
        const Duration(milliseconds: 500),
      );
      expect(
        delayForAttempt(1, jitter: zeroJitter),
        const Duration(milliseconds: 1000),
      );
      expect(
        delayForAttempt(2, jitter: zeroJitter),
        const Duration(milliseconds: 2000),
      );
      expect(
        delayForAttempt(3, jitter: zeroJitter),
        const Duration(milliseconds: 4000),
      );
      expect(
        delayForAttempt(4, jitter: zeroJitter),
        const Duration(milliseconds: 8000),
      );
    });

    test('plafonne à maxDelay / 2 dès la tentative 5, gigue nulle', () {
      expect(
        delayForAttempt(5, jitter: zeroJitter),
        const Duration(milliseconds: 15000),
      );
      expect(
        delayForAttempt(12, jitter: zeroJitter),
        const Duration(milliseconds: 15000),
      );
    });
  });

  group('delayForAttempt, gigue maximale (1.0)', () {
    double maxJitter() => 1.0;

    test('au plateau, la gigue double la base plafonnée jusqu\'à maxDelay', () {
      expect(
        delayForAttempt(5, jitter: maxJitter),
        const Duration(milliseconds: 30000),
      );
      expect(
        delayForAttempt(40, jitter: maxJitter),
        const Duration(milliseconds: 30000),
      );
    });

    test('avant le plateau, la gigue double la base', () {
      expect(
        delayForAttempt(0, jitter: maxJitter),
        const Duration(milliseconds: 1000),
      );
    });
  });

  group('delayForAttempt, gigue médiane (0.5)', () {
    double medianJitter() => 0.5;

    test('au plateau, la gigue reste visible : 22500 ms', () {
      expect(
        delayForAttempt(5, jitter: medianJitter),
        const Duration(milliseconds: 22500),
      );
      expect(
        delayForAttempt(20, jitter: medianJitter),
        const Duration(milliseconds: 22500),
      );
    });

    test('avant le plateau : 750 ms à la tentative 0', () {
      expect(
        delayForAttempt(0, jitter: medianJitter),
        const Duration(milliseconds: 750),
      );
    });
  });

  group('Bornes et erreurs', () {
    test('le délai ne dépasse jamais maxDelay, gigue maximale, 0 à 63', () {
      for (int attempt = 0; attempt <= 63; attempt++) {
        final Duration delay = delayForAttempt(attempt, jitter: () => 1.0);
        expect(
          delay <= maxDelay,
          isTrue,
          reason: 'tentative $attempt a produit $delay > $maxDelay',
        );
      }
    });

    test('une tentative négative lève', () {
      expect(() => delayForAttempt(-1), throwsArgumentError);
    });
  });
}
